sw/device/tests/otbn_mem_scramble_test.c

```diff
@@ -0,0 +1,101 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/lib/dif/dif_base.h"
+#include "sw/device/lib/dif/dif_otbn.h"
+#include "sw/device/lib/runtime/log.h"
+#include "sw/device/lib/runtime/otbn.h"
+#include "sw/device/lib/testing/check.h"
+#include "sw/device/lib/testing/test_framework/ottf.h"
+
+#include "hw/top_earlgrey/sw/autogen/top_earlgrey.h"
+
+OTBN_DECLARE_APP_SYMBOLS(randomness);
+static const otbn_app_t kOtbnAppCfiTest = OTBN_APP_T_INIT(randomness);
+
+const test_config_t kTestConfig;
+
+/**
+ * Check whether the content of an otbn memory match the reference data pointed
+ * by `addr`.
+ *
+ * @param ctx The otbn context object.
+ * @param addr The address to the reference data.
+ * @param mem_size The size to the reference data.
+ * @param read_function Pointer to the function to read the data. It can be
+ * either `dif_otbn_imem_read` or `dif_otbn_dmem_read`.
+ * @return The result of the operation.
+ */
+static otbn_result_t otbn_check_mem(
+    otbn_t *ctx, const uint8_t *addr, size_t mem_size,
+    dif_result_t (*read_function)(const dif_otbn_t *otbn, uint32_t offset_bytes,
+                                  void *dest, size_t len_bytes)) {
+  enum { kBufSize = 256 };
+  uint8_t local_buf[kBufSize] = {0};
+  size_t offset = 0;
+  do {
+    size_t remainder = (mem_size - offset);
+    size_t read_size = remainder < kBufSize ? remainder : kBufSize;
+    if (read_function(&ctx->dif, offset, local_buf, read_size) != kDifOk) {
+      return kOtbnError;
+    }
+
+    if (memcmp(addr + offset, local_buf, read_size)) {
+      return kOtbnError;
+    }
+    offset += read_size;
+  } while (offset < mem_size);
+  return kOtbnOk;
+}
+
+/**
+ * Check that the application is loaded correctly to the IMEM and DMEM.
+ *
+ * @param ctx The otbn context object.
+ * @param app The application to load into OTBN.
+ * @return The result of the operation.
+ */
+static otbn_result_t otbn_check_app(otbn_t *ctx, const otbn_app_t app) {
+  const size_t imem_size = app.imem_end - app.imem_start;
+  const size_t data_size = app.dmem_data_end - app.dmem_data_start;
+
+  // Memory images and offsets must be multiples of 32b words.
+  if (imem_size % sizeof(uint32_t) != 0) {
```

Comment:
This should also be a CHECK. Prefer to CHECK as locally as possible in tests, since it provides better failure output.